### ISSUE TITLE
vkMapMemory and vkBindImageMemory call ordering issue

### ISSUE THREAD

**Issue by ch45er** [KhronosGroup/MoltenVK#163]
Hi,

I am observing a problem while uploading image data to the GPU.  My case is as simple as to create a linearly-tiled image with host-writable and host-coherent backing memory; then to copy the pixel data and to draw some textured geometry. Pretty much the same as the well-known LunarG cube demo does.

If I do the work in the following order:
1. vkCreateImage
2. vkAllocateMemory
3. vkBindImageMemory
4. vkMapMemory + copy + vkUnmapMemory

everything works as expected. However, if I swap calls 3 and 4, so that the `map+copy+unmap` sequence appears *before* vkBindImageMemory, the result is completely black image.

I've tested both ways on a Windows+GTX1070 machine with 'native' Vulkan and had no issues whatsoever. According to the specs, there seems to be no ordering requirements for these calls. 

This [gist](https://gist.github.com/ch45er/830e4baf96c7cb88f87c54beb801c9f9) illustrates the change and may be used to reproduce the issue using the cube demo. Points of interest are marked with `POI` in comments.

This doesn't seem to happen with buffers, except texel buffers, which are apparently backed with images in MoltenVK.

I am using MoltenVK ver. 1.0.3 (bundled with LunarG SDK (vulkansdk-macos-1.1.73.0)), macOS 10.13.4. Hardware: MacBook Pro (Mid 2015) with Intel Iris Pro.

An easy way to encounter such a problem is to allocate resource memory via [VulkanMemoryAllocator](https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator) by calling vmaCreateImage and passing VmaAllocationCreateInfo with VMA_ALLOCATION_CREATE_MAPPED_BIT flag. This will cause the memory to be mapped before it is bound to the VkImage.

**Comment by billhollings** [comment]
@ch45er 

This has been fixed in PR #173. Please verify and close this issue if successful.

**Comment by ch45er** [comment]
Hi @billhollings,

with the patches applied, the modfied cube demo works properly. However, I am still experiencing (seemingly the same) problem with the VMA (as described above). I would like to take some time to debug it and make sure it's (un)related.

Thanks!

**Comment by ch45er** [comment]
@billhollings 

I have an update. What is happening now:
* vkAllocateMemory
* vkBind*Memory
* vkMapMemory + copy + vkUnmapMemory

sequence works no matter if `vkBind*Memory` appears before or after `vkUnmapMemory`. 

But the underlying Metal resource update actually happens upon `vkBind*Memory` or `vkUnmapMemory`.

In my case, the VkDeviceMemory has `VK_MEMORY_PROPERTY_HOST_COHERENT_BIT` set and is intended to be permanently mapped. The memory is used for resource uploading and staging resources are sub-allocated from it. Since the memory is host coherent, I expect any writes to that memory to be instantly visible to the forthcoming commands (e.g. vkCmdCopyImage) as the spec says: "the host cache management commands [...] are not needed to flush host writes to the device".

What I am seeing however, is that synchronization is required with MoltenVK.

In the modified cube demo case, the writes are implicitly flushed by either `vkBind*Memory` or `vkUnmapMemory`. For example, if you "forget" to unmap the memory in `demo_prepare_texture_image`, you will see a completely black texture. This may be "fixed" though, by moving `vkBindImageMemory` down (to the second POI mark).

Should I open a new issue for this? Is host coherency in Vulkan terms even possible in Metal?

**Comment by billhollings** [comment]
@ch45er 

Unfortunately...you are correct in your last statement. Host-coherent texture memory is not natively supported under Metal.

However...it might be possible to deal with this if we limit mapped memory to linear textures that only support transfer operations (as per the Vulkan spec)...and then implement such a `VkImage` as a `MTLBuffer` instead of as a `MTLTexture`. A `MTLBuffer` can be coherently mapped via a permanent `vkMapMemory()` call.

I have opened a new issue #175 as a separate enhancement request for that...and I am closing this issue...since your original problem has been resolved.
